sdks/java/io/common/src/main/java/org/apache/beam/sdk/io/common/synthetic/SyntheticBoundedInput.java

```diff
@@ -0,0 +1,452 @@
+     * defined by {@link #initializeDelayDistribution}.
+     */
+    public Duration nextInitializeDelay(long seed) {
+      return Duration.millis((long) initializeDelayDistribution.sample(seed));
+    }
+
+    @Override
+    public void validate() {
+      super.validate();
+      checkArgument(
+          numRecords >= 0, "numRecords should be a non-negative number, but found %s.", numRecords);
+      checkNotNull(bundleSizeDistribution, "bundleSizeDistribution");
+      checkArgument(
+          forceNumInitialBundles == null || forceNumInitialBundles > 0,
+          "forceNumInitialBundles, if specified, must be positive, but found %s",
+          forceNumInitialBundles);
+      checkArgument(
+          splitPointFrequencyRecords >= 0,
+          "splitPointFrequencyRecords must be non-negative, but found %s",
+          splitPointFrequencyRecords);
+    }
+
+    public Record genRecord(long position) {
+      // This method is supposed to generate random records deterministically,
+      // so that results can be reproduced by running the same scenario a second time.
+      // We need to initiate a Random object for each position to make the record deterministic
+      // because liquid sharding could split the Source at any position.
+      // And we also need a seed to initiate a Random object. The mapping from the position to
+      // the seed should be fixed. Using the position as seed to feed Random objects will cause the
+      // generated values to not be random enough because the position values are
+      // close to each other. To make seeds fed into the Random objects unrelated,
+      // we use a hashing function to map the position to its corresponding hashcode,
+      // and use the hashcode as a seed to feed into the Random object.
+      long hashCodeOfPosition = hashFunction.hashLong(position).asLong();
+      return new Record(genKvPair(hashCodeOfPosition), nextDelay(hashCodeOfPosition));
+    }
+
+    /**
+     * Record generated by {@link #genRecord}.
+     */
+    public static class Record {
+      public final KV<byte[], byte[]> kv;
+      public final Duration sleepMsec;
+
+      public Record(KV<byte[], byte[]> kv, long sleepMsec) {
+        this.kv = kv;
+        this.sleepMsec = new Duration(sleepMsec);
+      }
+    }
+  }
+
+  /**
+   * A reader over the {@link PCollection} of {@code KV<byte[], byte[]>}
+   * from the synthetic source.
+   *
+   * <p>The random but deterministic record at position "i" in the range [A, B)
+   * is generated by using {@link SourceOptions#genRecord}.
+   * Reading each record sleeps according to the sleep time distribution
+   * in {@code SyntheticOptions}.
+   */
+  private static class SyntheticSourceReader
+      extends OffsetBasedSource.OffsetBasedReader<KV<byte[], byte[]>> {
+    private final long splitPointFrequencyRecords;
+
+    private KV<byte[], byte[]> currentKvPair;
+    private long index;
```

Comment:
probably more clear to call this one `currentOffset` to be consistent with all the use of Offset and the get method.